File: shared/pomodoro.js
// Gemeinsame, reine Pomodoro-Domänenlogik.
// Läuft unverändert in Browser (PWA), Node-Backend und Chrome-Extension.
// Kein DOM, kein Storage, kein Side-Effect — nur Zustandsübergänge & Zeit-Mathematik.

export const PHASES = Object.freeze({
  FOCUS: "focus",
  SHORT_BREAK: "short-break",
  LONG_BREAK: "long-break",
});

export const STATUS = Object.freeze({
  IDLE: "idle",
  RUNNING: "running",
  PAUSED: "paused",
});

export const DEFAULT_SETTINGS = Object.freeze({
  focusMinutes: 25,
  shortBreakMinutes: 5,
  longBreakMinutes: 15,
  cyclesUntilLongBreak: 4,
  autoStartNextPhase: false,
  todayGoalHours: 4,
  profileName: "Prüfungsfokus",
});

// ── Zeit-Helfer ──────────────────────────────────
export function minutesToMs(minutes) {
  return Math.max(1, Number(minutes)) * 60 * 1000;
}

export function getPhaseDurationMs(phase, settings) {
  if (phase === PHASES.FOCUS) return minutesToMs(settings.focusMinutes);
  if (phase === PHASES.SHORT_BREAK) return minutesToMs(settings.shortBreakMinutes);
  return minutesToMs(settings.longBreakMinutes);
}

export function formatMs(value) {
  const seconds = Math.ceil(Math.max(0, value) / 1000);
  const minutes = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return `${String(minutes).padStart(2, "0")}:${String(secs).padStart(2, "0")}`;
}

// Dauer OHNE Sekunden – "nur h und min". "25 Min", "1 Std 05 Min", "2 Std".
// ceil=true für laufende Countdowns (bleibt bis zur vollen Minute stehen, statt
// vorzeitig herunterzuspringen); ceil=false (Default) für verstrichene Dauern.
export function formatDurationHM(value, ceil = false) {
  const ms = Math.max(0, value);
  const totalMin = ceil ? Math.ceil(ms / 60000) : Math.round(ms / 60000);
  const h = Math.floor(totalMin / 60);
  const m = totalMin % 60;
  if (h && m) return `${h} Std ${String(m).padStart(2, "0")} Min`;
  if (h) return `${h} Std`;
  return `${m} Min`;
}

// ── Settings-Validierung ─────────────────────────
function clampInt(value, min, max, fallback) {
  const number = Number(value);
  if (!Number.isFinite(number)) return fallback;
  return Math.min(max, Math.max(min, Math.round(number)));
}

function clampFloat(value, min, max, fallback) {
  const number = Number(value);
  if (!Number.isFinite(number)) return fallback;
  return Math.min(max, Math.max(min, number));
}

export function sanitizeSettings(raw) {
  const base = { ...DEFAULT_SETTINGS, ...(raw || {}) };
  return {
    focusMinutes: clampInt(base.focusMinutes, 5, 90, DEFAULT_SETTINGS.focusMinutes),
    shortBreakMinutes: clampInt(base.shortBreakMinutes, 1, 30, DEFAULT_SETTINGS.shortBreakMinutes),
    longBreakMinutes: clampInt(base.longBreakMinutes, 5, 45, DEFAULT_SETTINGS.longBreakMinutes),
    cyclesUntilLongBreak: clampInt(base.cyclesUntilLongBreak, 2, 8, DEFAULT_SETTINGS.cyclesUntilLongBreak),
    autoStartNextPhase: Boolean(base.autoStartNextPhase),
    todayGoalHours: clampFloat(base.todayGoalHours, 0.5, 16, DEFAULT_SETTINGS.todayGoalHours),
    profileName:
      typeof base.profileName === "string" && base.profileName.trim()
        ? base.profileName.trim()
        : DEFAULT_SETTINGS.profileName,
  };
}

// ── Zustands-Erzeugung ───────────────────────────
export function createInitialState(settings = DEFAULT_SETTINGS, now = Date.now()) {
  return {
    status: STATUS.IDLE,
    phase: PHASES.FOCUS,
    cycleInBlock: 0,
    remainingMs: getPhaseDurationMs(PHASES.FOCUS, settings),
    endsAt: null,
    activeTaskId: null,
    phaseStartedAt: null,
    updatedAt: now,
  };
}

export function computeRemainingMs(state, now = Date.now()) {
  if (state.status !== STATUS.RUNNING || !state.endsAt) return state.remainingMs;
  return Math.max(0, state.endsAt - now);
}

// ── Übergänge (immutable) ────────────────────────
export function startPhase(state, settings, now = Date.now()) {
  if (state.status === STATUS.RUNNING) return state;
  const remainingMs = Math.max(1000, state.remainingMs || getPhaseDurationMs(state.phase, settings));
  return {
    ...state,
    status: STATUS.RUNNING,
    remainingMs,
    endsAt: now + remainingMs,
    phaseStartedAt: now,
    updatedAt: now,
  };
}

export function pausePhase(state, now = Date.now()) {
  if (state.status !== STATUS.RUNNING) return state;
  return {
    ...state,
    status: STATUS.PAUSED,
    remainingMs: computeRemainingMs(state, now),
    endsAt: null,
    phaseStartedAt: null,
    updatedAt: now,
  };
}

export function resumePhase(state, settings, now = Date.now()) {
  if (state.status !== STATUS.PAUSED) return state;
  const remainingMs = Math.max(1000, state.remainingMs || getPhaseDurationMs(state.phase, settings));
  return {
    ...state,
    status: STATUS.RUNNING,
    remainingMs,
    endsAt: now + remainingMs,
    phaseStartedAt: now,
    updatedAt: now,
  };
}

export function resetSession(state, settings, now = Date.now()) {
  return {
    ...createInitialState(settings, now),
    // aktive Aufgabe bleibt beim Reset erhalten
    activeTaskId: state?.activeTaskId ?? null,
  };
}

export function advanceToNextPhase(state, settings, now = Date.now()) {
  const fromPhase = state.phase;
  let nextPhase = PHASES.FOCUS;
  let cycleInBlock = state.cycleInBlock;

  if (fromPhase === PHASES.FOCUS) {
    const progressed = cycleInBlock + 1;
    if (progressed >= settings.cyclesUntilLongBreak) {
      nextPhase = PHASES.LONG_BREAK;
      cycleInBlock = 0;
    } else {
      nextPhase = PHASES.SHORT_BREAK;
      cycleInBlock = progressed;
    }
  } else {
    nextPhase = PHASES.FOCUS;
  }

  const remainingMs = getPhaseDurationMs(nextPhase, settings);
  const autoStart = Boolean(settings.autoStartNextPhase);
  return {
    ...state,
    status: autoStart ? STATUS.RUNNING : STATUS.PAUSED,
    phase: nextPhase,
    cycleInBlock,
    remainingMs,
    endsAt: autoStart ? now + remainingMs : null,
    phaseStartedAt: autoStart ? now : null,
    updatedAt: now,
  };
}

// Phase manuell wählen (Tabs im UI). Setzt Timer auf die Dauer der Phase zurück.
export function selectPhase(state, settings, phase, now = Date.now()) {
  if (![PHASES.FOCUS, PHASES.SHORT_BREAK, PHASES.LONG_BREAK].includes(phase)) return state;
  return {
    ...state,
    status: STATUS.IDLE,
    phase,
    remainingMs: getPhaseDurationMs(phase, settings),
    endsAt: null,
    phaseStartedAt: null,
    updatedAt: now,
  };
}

// Wird nach Settings-Änderung aufgerufen: passt die Restzeit an, wenn nicht läuft.
export function applySettings(state, settings, now = Date.now()) {
  if (state.status === STATUS.RUNNING) {
    return { ...state, updatedAt: now };
  }
  return {
    ...state,
    remainingMs: getPhaseDurationMs(state.phase, settings),
    endsAt: null,
    updatedAt: now,
  };
}

// Priorität P1 (dringend) … P4 (kann warten). Label-Helfer, geteilt vom UI.
export const PRIORITY = Object.freeze({ P1: 1, P2: 2, P3: 3, P4: 4 });

export function phaseLabelJa(phase) {
  if (phase === PHASES.FOCUS) return "Focus";
  if (phase === PHASES.SHORT_BREAK) return "Short break";
  return "Long break";
}

export function phaseLabelDe(phase) {
  if (phase === PHASES.FOCUS) return "Fokus";
  if (phase === PHASES.SHORT_BREAK) return "Kurze Pause";
  return "Lange Pause";
}

File: Pace/Sources/PaceKit/Domain/Pomodoro.swift
import Foundation

// MARK: - Focus engine (ported from src/domain/pomodoroDomain.js — pure, immutable state)

public enum Phase: String, Sendable, Equatable {
    case focus, shortBreak, longBreak
}

public enum TimerStatus: String, Sendable, Equatable {
    case idle, running, paused
}

public struct PomodoroSettings: Sendable, Equatable {
    public var focusMinutes: Int
    public var shortBreakMinutes: Int
    public var longBreakMinutes: Int
    public var cyclesUntilLongBreak: Int
    public var autoStartNextPhase: Bool

    public init(focusMinutes: Int = 25, shortBreakMinutes: Int = 5, longBreakMinutes: Int = 15, cyclesUntilLongBreak: Int = 4, autoStartNextPhase: Bool = false) {
        self.focusMinutes = focusMinutes
        self.shortBreakMinutes = shortBreakMinutes
        self.longBreakMinutes = longBreakMinutes
        self.cyclesUntilLongBreak = cyclesUntilLongBreak
        self.autoStartNextPhase = autoStartNextPhase
    }

    public static let `default` = PomodoroSettings()

    public func sanitized() -> PomodoroSettings {
        PomodoroSettings(
            focusMinutes: clampInt(focusMinutes, 10, 90, 25),
            shortBreakMinutes: clampInt(shortBreakMinutes, 3, 30, 5),
            longBreakMinutes: clampInt(longBreakMinutes, 10, 45, 15),
            cyclesUntilLongBreak: clampInt(cyclesUntilLongBreak, 2, 6, 4),
            autoStartNextPhase: autoStartNextPhase
        )
    }
}

public struct PomodoroState: Sendable, Equatable {
    public var status: TimerStatus
    public var phase: Phase
    public var cycleInBlock: Int
    public var remaining: TimeInterval   // seconds left in the current phase
    public var endsAt: Date?
    public var updatedAt: Date

    public init(status: TimerStatus, phase: Phase, cycleInBlock: Int, remaining: TimeInterval, endsAt: Date?, updatedAt: Date) {
        self.status = status
        self.phase = phase
        self.cycleInBlock = cycleInBlock
        self.remaining = remaining
        self.endsAt = endsAt
        self.updatedAt = updatedAt
    }
}

/// Pure state transitions — same logic as the JS domain, expressed as value-returning functions.
public enum Pomodoro {

    public static func phaseDuration(_ phase: Phase, _ s: PomodoroSettings) -> TimeInterval {
        switch phase {
        case .focus:      return TimeInterval(max(1, s.focusMinutes) * 60)
        case .shortBreak: return TimeInterval(max(1, s.shortBreakMinutes) * 60)
        case .longBreak:  return TimeInterval(max(1, s.longBreakMinutes) * 60)
        }
    }

    public static func initialState(_ s: PomodoroSettings, now: Date = Date()) -> PomodoroState {
        PomodoroState(status: .idle, phase: .focus, cycleInBlock: 0,
                      remaining: phaseDuration(.focus, s), endsAt: nil, updatedAt: now)
    }

    public static func computeRemaining(_ state: PomodoroState, now: Date) -> TimeInterval {
        guard state.status == .running, let endsAt = state.endsAt else { return state.remaining }
        return max(0, endsAt.timeIntervalSince(now))
    }

    public static func start(_ state: PomodoroState, _ s: PomodoroSettings, now: Date = Date()) -> PomodoroState {
        let remaining = state.remaining > 0 ? state.remaining : phaseDuration(state.phase, s)
        var ns = state
        ns.status = .running
        ns.remaining = remaining
        ns.endsAt = now.addingTimeInterval(remaining)
        ns.updatedAt = now
        return ns
    }

    public static func pause(_ state: PomodoroState, now: Date = Date()) -> PomodoroState {
        guard state.status == .running else { return state }
        var ns = state
        ns.remaining = computeRemaining(state, now: now)
        ns.status = .paused
        ns.endsAt = nil
        ns.updatedAt = now
        return ns
    }

    public static func resume(_ state: PomodoroState, _ s: PomodoroSettings, now: Date = Date()) -> PomodoroState {
        guard state.status == .paused else { return state }
        return start(state, s, now: now)
    }

    public static func reset(_ s: PomodoroSettings, now: Date = Date()) -> PomodoroState {
        var st = initialState(s, now: now)
        st.updatedAt = now
        return st
    }

    public static func advance(_ state: PomodoroState, _ s: PomodoroSettings, now: Date = Date()) -> PomodoroState {
        var nextPhase: Phase = .focus
        var cycle = state.cycleInBlock

        if state.phase == .focus {
            let progressed = cycle + 1
            if progressed >= s.cyclesUntilLongBreak {
                nextPhase = .longBreak
                cycle = 0
            } else {
                nextPhase = .shortBreak
                cycle = progressed
            }
        } else {
            nextPhase = .focus
        }

        let dur = phaseDuration(nextPhase, s)
        var ns = state
        ns.status = s.autoStartNextPhase ? .running : .paused
        ns.phase = nextPhase
        ns.cycleInBlock = cycle
        ns.remaining = dur
        ns.endsAt = s.autoStartNextPhase ? now.addingTimeInterval(dur) : nil
        ns.updatedAt = now
        return ns
    }
}

fileprivate func clampInt(_ v: Int, _ lo: Int, _ hi: Int, _ fallback: Int) -> Int {
    if v < lo { return lo }
    if v > hi { return hi }
    return v
}
